tools/cachestat_example.txt

```diff
@@ -15,61 +15,42 @@ examples:
-16:07:16   305701        0        0   100.0%     0.0%            0      11336
-16:07:17   312754        0        0   100.0%     0.0%            0      11336
-16:07:18   308406        0        0   100.0%     0.0%            0      11336
-16:07:19   298185        0        0   100.0%     0.0%            0      11336
-16:07:20   236128        0        0   100.0%     0.0%            0      11336
-16:07:21   257616        0        0   100.0%     0.0%            0      11336
-16:07:22   179792        0        0   100.0%     0.0%            0      11336
-
-Command used to generate the activity
-# dd if=/root/mnt2/testfile of=/dev/null bs=8192
-1442795+0 records in
-1442795+0 records out
-11819376640 bytes (12 GB) copied, 3.9301 s, 3.0 GB/s
-
-Below shows the dirty ratio increasing as we add a file to the page cache 
-# ./cachestat.py
-    HITS   MISSES  DIRTIES  READ_HIT% WRITE_HIT%   BUFFERS_MB  CACHED_MB
-    1074       44       13      94.9%       2.9%            1        223
-    2195      170        8      92.5%       6.8%            1        143
-     182       53       56      53.6%       1.3%            1        143
-   62480    40960    20480      40.6%      19.8%            1        223
-       7        2        5      22.2%      22.2%            1        223
-     348        0        0     100.0%       0.0%            1        223
-       0        0        0       0.0%       0.0%            1        223
-       0        0        0       0.0%       0.0%            1        223
-       0        0        0       0.0%       0.0%            1        223
-       0        0        0       0.0%       0.0%            1        223
-
-The file copied into page cache was named 80m with a size of 83886080 (83886080/4096) = 20480, this matches what we see under dirties
+Following commands show a 2GB file being read into the page cache.
+
+Command used to generate activity:
+# dd if=/root/tmpfile of=/dev/null bs=8192
+
+Output from cachestat running simultatenously:
+# ./tools/cachestat.py 1
+   TOTAL   MISSES     HITS  DIRTIES   BUFFERS_MB  CACHED_MB
+       1        0        1        0            8        283
+       0        0        0        0            8        283
+       0        0        0        2            8        283
+       0        0        0        0            8        283
+   10009     9173      836        2            9        369
+  152032   152032        0        0            9       1028
+  157408   157405        3        0            9       1707
+  150432   150432        0        0            9       2331
+       0        0        0        0            9       2331
+       1        1        0        1            9       2331
+       0        0        0        0            9       2331
+       0        0        0        0            9       2331
+       0        0        0        0            9       2331
+
+The misses counter reflects a 2GB file being read and almost everything being
```

Comment:
I actually run the command with a 20MB file on latest bpf-next. With debug = 1, I got:
```
7549 0 508 700 7549 0 7549
mpa, mbd, apcl, apd, total, misses, hits
```
Here, `apcl` is less then `apd`.
Look at the kernel code, it is not clear to me what is the relationship between
`add_to_page_cache_lru` and `account_page_dirtied`. `

Reply:
`add_to_page_cache_lru` counts pages added to the page cache which included both pages added because of read misses and pages added because of writes. IMO a "write miss" isn't meaningful since the page will be lazy written back to the disk and isn't a harmful miss, so such "adds" to the page cache should not be counted, and this is the reason why  `account_page_dirtied` is subtracted.

Here's a run of funccount counting all these, during some dd's:
```
# DD commands:
dd if=/dev/urandom of=tmp2 bs=4k count=$((256 * 20))
dd if=tmp1 of=/dev/null bs=4k
```
```
# ./tools/funccount.py -r  '(add_to_page_cache_lru|account_page_dirtied|mark_buffer_dirty|mark_page_accessed)'
Tracing 5 functions for "(add_to_page_cache_lru|account_page_dirtied|mark_buffer_dirty|mark_page_accessed)"... Hit Ctrl-C to end.
^C
FUNC                                    COUNT
account_page_dirtied                     5120
mark_buffer_dirty                        5120
add_to_page_cache_lru                   13153
mark_page_accessed                      16103
```

apcl - abd = ~8k pages which is the 30MB file that was added to the page cache as a part of the "read" dd and these are the hurtful misses.

Unfortunately, I think its possible a page already exists in the page cache and is being dirtied, so these `account_page_dirtied` are the extra ones which don't have a corresponding `add_to_page_cache_lru`. 

Another thing to note is Brendan's script also assumed `apcl` can be < `apd`: https://github.com/brendangregg/perf-tools/blob/master/fs/cachestat#L150

Reply:
Just to clarify, the `tmp1` file above is the 30MB I had dd'd in advance.

Reply:
Do you think it makes sense to have finer control on add_to_page_cache_lru/account_page_dirtied? For example, have a map shared between these two functions and account_page_dirtied event is counted if only page has been added to lru?

Reply:
Yeah I think so, but I think it will also complicate the script more. I am wondering how often this negative value happens for you? I believe the script is expected to provide "rough estimate" than accurate results. And also that, in the future the meanings and semantics can change anyway so I am wondering if its worth handling this case (unless it happens way too often)